File: gherkin/dotnet/Gherkin/Stream/GherkinEvents.cs
﻿using Gherkin.Events;
using Gherkin.Pickles;
using System;
using System.Collections.Generic;
using System.IO;
using Gherkin.CucumberMessages.Types;
using Gherkin.Stream.Converter;
using GherkinDocument = Gherkin.Ast.GherkinDocument;
using Location = Gherkin.CucumberMessages.Types.Location;

namespace Gherkin.Stream
{
    public class GherkinEvents
    {
        private readonly Parser _parser = new Parser();
        private readonly Compiler _compiler = new Compiler();
        private readonly AstEventConverter _astEventConverter = new AstEventConverter();
        private readonly SourceEventConverter _sourceEventConverter = new SourceEventConverter();

        readonly bool _printAst;
        readonly bool _printPickles;
        readonly bool _printSource;

        public GherkinEvents(bool printSource, bool printAst, bool printPickles)
        {
            _printSource = printSource;
            _printAst = printAst;
            _printPickles = printPickles;
        }

        public IEnumerable<IEvent> Iterable(Sources sourceEvent)
        {
            List<IEvent> events = new List<IEvent>();

            try
            {
                GherkinDocument gherkinDocument = _parser.Parse(new StringReader(sourceEvent.Data));

                if (_printSource)
                {
                    events.Add(_sourceEventConverter.Convert(sourceEvent));
                }
                if (_printAst)
                {
                    events.Add(new GherkinDocumentEvent(_astEventConverter.ConvertGherkinDocumentToEventArgs(gherkinDocument, sourceEvent.Uri)));
                }
                if (_printPickles)
                {
                    List<Pickle> pickles = _compiler.Compile(_astEventConverter.ConvertGherkinDocumentToEventArgs(gherkinDocument, sourceEvent.Uri));
                    foreach (Pickle pickle in pickles)
                    {
                        events.Add(new PickleEvent(pickle));
                    }
                }
            }
            catch (CompositeParserException e)
            {
                foreach (ParserException error in e.Errors)
                {
                    addErrorAttachment(events, error, sourceEvent.Uri);
                }
            }
            catch (ParserException e)
            {
                addErrorAttachment(events, e, sourceEvent.Uri);
            }
            return events;
        }


        private void addErrorAttachment(List<IEvent> events, ParserException e, String uri)
        {
            events.Add(new ParseErrorEvent()
            {
                EventArgs = new ParseError()
                {
                    Message = e.Message,
                    Source = new SourceReference()
                    {
                        Location = new Location(e.Location.Column, e.Location.Line),
                        Uri = uri
                    }
                }
            });

        }
    }
}

File: gherkin/python/gherkin/stream/gherkin_events.py
from gherkin.ast_builder import AstBuilder
from gherkin.parser import Parser
from gherkin.pickles.compiler import Compiler
from gherkin.errors import ParserError, CompositeParserException
from gherkin.stream.id_generator import IdGenerator

def create_errors(errors, uri):
    for error in errors:
        yield {
            'parseError': {
            'source': {
                'uri': uri,
                'location': error.location
            },
            'message': str(error),
        }}


class GherkinEvents:
    def __init__(self, options):
        self.options = options
        self.id_generator = IdGenerator()
        self.parser = Parser(ast_builder=AstBuilder(self.id_generator))
        self.compiler = Compiler(self.id_generator)

    def enum(self, source_event):
        uri = source_event['source']['uri']
        source = source_event['source']['data']

        try:
            gherkin_document = self.parser.parse(source)
            gherkin_document['uri'] = uri

            if (self.options.print_source):
                yield source_event

            if (self.options.print_ast):
                yield {
                    'gherkinDocument': gherkin_document
                }

            if (self.options.print_pickles):
                pickles = self.compiler.compile(gherkin_document)
                for pickle in pickles:
                    yield {
                        'pickle': pickle
                    }
        except CompositeParserException as e:
            for event in create_errors(e.errors, uri):
                yield event
        except ParserError as e:
            for event in create_errors([e], uri):
                yield event
